docs/website/contents/for-developers/HardWonWisdom.md

```diff
@@ -348,3 +348,77 @@ The key corresponding parts of the implementation are as follows.
 - The interpretation only depends on the eras' start and end bounds, not on the safe zone.
   The safe zone is instead used in an preceding step to safely determine the least possible value of the first unknown end bound.
   This determination is implemented by the `ouroboros-consensus:Ouroboros.Consensus.HardFork.Combinator.State.reconstructSummaryLedger` function.
+
+## The hard fork combinator interface
+
+To run the Consensus Layer on a block type `blk` requires the following (whether or not it's a `HardForkBlock`).
+
+- Values for various tracers, ChainDB options, etc, as required by `Ouroboros.Consensus.Node.run`.
+- A `ProtocolInfo blk` value, which is `TopLevelConfig blk` as well as initial state of the ledger and protocol.
+- Instances of `ConsensusProtocol (BlockProtocol blk)`, `LedgerSupportsProtocol blk`, and so on (see the `RunNode blk` class).
+
+Suppose there are two blocks types `F` and `G` for which the above is already available.
+To run the Consensus Layer instead on the hard fork combinator that sequences `F` into `G` requires the following additional information about `F` and `G`.
+
+- A value of `EraParams` for `F` and for `G` (ie `Shape [F, G]`), which decomposes as follows.
+    - The duration of every slot in the epoch as a [`NominalDiffTime`](https://hackage.haskell.org/package/time-1.12.2/docs/Data-Time-Clock.html#t:NominalDiffTime).
+    - The duration of every epoch in the era as a number of slots.
+    - The era's _safe zone_, which depends on the `singleEraTransition` function; see below.
+        - If the era's safe zone is `d`, then `singleEraTransition` cannot return `Just e` for the ledger state resulting from some block `x` (possibly the "genesis block") (possibly from an earlier era, and not even necessarily the previous era) unless there are at least `d` slots between `x` and `e`, ie `slot(x) + d < firstSlot(e)`.
+        - The fundamental motivation for the HFC combinator is to enable a chain to eventualy seemlessly transition into eras that do not yet exist.
```

Comment:
```suggestion
        - The fundamental motivation for the HFC combinator is to enable a chain to eventually seamlessly transition into eras that do not yet exist.
```
Also, is this actually *the* fundamental motivation and not just *one* of them (eg the HFC would still be very useful even if we knew that there won't be any era transitions in the foreseeable future)?

Reply:
Well, the HFC was created in the first place in order to solve that problem for Byron-to-Shelley.

Reply:
Right, then maybe it was just unclear to me what an era "existing" means, ie you could say that during the development preparing for the Byron-to-Shelley HF, Shelley did already exist (and was fully implemented), but no block in the Shelley era was yet on mainnet. In contrast, at that time, Conway didn't even exist as a concept. Please ignore if you think this is just hairsplitting.

Reply:
> eg the HFC would still be very useful even if we knew that there won't be any era transitions in the foreseeable future

What's the use case you're thinking of here? Specifically, why are there multiple separately-implemented eras to being with? Perhaps just for re-use, I guess?

Reply:
Suppose that eg in some future era, everyone is super happy with Cardano and there is no majority that is willing to fund work on a new era that might only bring marginal improvements. Then the HFC is still very useful to allow syncing to this chain from Genesis.

But that probably only occurs to me because I can look back now at several eras; I agree that it definitely was the fundamental motivation for the HFC back then :+1: